File: src/main/java/org/gd/leetcode/p0124/Solution.java
package org.gd.leetcode.p0124;

import org.gd.common.Repeat;
import org.gd.leetcode.common.LeetCode;
import org.gd.leetcode.common.TreeNode;

/**
 * https://leetcode.com/problems/binary-tree-maximum-path-sum/
 */
@Repeat("longest path in binary tree")
@LeetCode(
        difficulty = LeetCode.Level.HARD,
        tags = {
                LeetCode.Tags.TREE,
                LeetCode.Tags.DEPTH_FIRST_SEARCH
        })
class Solution {

    private int max;

    private int calcPath(TreeNode root) {
        if (root == null)
            return 0;

        int leftPath = Math.max(0, calcPath(root.left));
        int rightPAth = Math.max(0, calcPath(root.right));

        max = Math.max(max, root.val + leftPath + rightPAth);

        return Math.max(leftPath, rightPAth) + root.val;
    }

    public int maxPathSum(TreeNode root) {
        max = Integer.MIN_VALUE;
        calcPath(root);
        return max;
    }
}

File: src/main/java/org/gd/leetcode/p0124/solution.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def __init__(self):
        self.__max_val = 0

    def maxPathSum(self, root: TreeNode) -> int:
        self.__max_val = -(1 << 31)

        def calc_path(node: TreeNode) -> int:
            if not node:
                return 0

            left_path = max(0, calc_path(node.left))
            rigth_path = max(0, calc_path(node.right))

            self.__max_val = max(self.__max_val, left_path + rigth_path + node.val)

            return max(left_path, rigth_path) + node.val

        calc_path(root)
        return self.__max_val
